LoopKit/InsulinKit/DoseStore.swift

```diff
@@ -1502,6 +1502,84 @@ extension DoseStore {
 
 }
 
+// MARK: - Critical Event Log Export
+
+extension DoseStore: CriticalEventLog {
+    private var exportEstimatedDurationPerObject: TimeInterval { 0.0005 }   // Estimated during development on iPhone 8, absolute duration is less important than relative to other critical event logs
+    private var exportFetchLimit: Int { Int(criticalEventLogExportMinimumProgressDuration / exportEstimatedDurationPerObject) }
+
+    public var exportName: String { "Doses.json" }
+
+    public func exportEstimatedDuration(startDate: Date, endDate: Date? = nil) -> Result<TimeInterval, Error> {
+        var result: Result<TimeInterval, Error>?
+
+        self.persistenceController.managedObjectContext.performAndWait {
+            do {
+                let request: NSFetchRequest<PumpEvent> = PumpEvent.fetchRequest()
+                request.predicate = self.exportDatePredicate(startDate: startDate, endDate: endDate)
+
+                let objectCount = try self.persistenceController.managedObjectContext.count(for: request)
+                result = .success(Double(objectCount) * exportEstimatedDurationPerObject)
+            } catch let error {
+                result = .failure(error)
+            }
+        }
+
+        return result!
+    }
+
+    public func export(startDate: Date, endDate: Date, to stream: OutputStream, progressor: EstimatedDurationProgressor) -> Error? {
```

Comment:
These `export(...)` functions on the Stores are so similar.  Is there a way to factor them into common code?

Reply:
I gave it a try and found there are too many differences between the different stores for it to be straightforward. It is possible, of course, but there would have to be at least on "customizer" function to account for the store differences and the final code would be less understandable. 

For example, each store counts each object as a different amount towards overall progress (which is because each object takes a different amount of time to query, serialize, and write - there is a very large difference in object size between, say, alerts and dosing decisions). So each store would have to customize this cost while the common code is iterating over and exporting each object.

Also, while most stores use a `modificationCounter` property to sort and query on, the CarbStore uses an `anchorKey` property. These properties are used to provide a stable sort for paging data during an export.

These differences and others would make the code much less understandable if we tried to make a common "export" function.

Long term we could work on making the stores more similar in functionality (and property names) and this might be easier.